presto-main/src/main/java/com/facebook/presto/execution/scheduler/CTEMaterializationTracker.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.execution.scheduler;
+
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.SettableFuture;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/*
+ *    Tracks whether tablefinish nodes writing temporary tables for CTE Materialization are complete
+ */
+@ThreadSafe
+public class CTEMaterializationTracker
+{
+    private final Map<String, SettableFuture<Void>> materializationFutures = new ConcurrentHashMap<>();
+
+    private final Map<String, Boolean> materializedCtes = new ConcurrentHashMap<>();
+
+    public ListenableFuture<Void> getFutureForCTE(String cteName)
+    {
+        return materializationFutures.compute(cteName, (key, existingFuture) -> {
+            if (existingFuture == null || existingFuture.isCancelled()) {
```

Comment:
why do we change isCancelled? Seems like that could be a race condition anyway (gets cancelled by another thread after we check the condition).

Reply:
As we know, the Scheduler’s [schedule()](https://github.com/prestodb/presto/commit/baf737b12240068a54aef16a2990ee624821519e) method is single-threaded. When the CTE writer stage (or another blocked stage) becomes unblocked [here](https://github.com/prestodb/presto/blob/85cc36af2cca5be4b15d17884f3f6a01bd107781/presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java#L514), the schedule() method resumes and cancels all pending futures [here](https://github.com/prestodb/presto/blob/85cc36af2cca5be4b15d17884f3f6a01bd107781/presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java#L517). Then it will try to schedle all blocked stages again and we will call FixedSourcePartitionedScheduler’s schedule() method, at which point a cancelled future will appear in the CTEMaterializationMap.

I think you are correct when you say that a race condition can appear in some extremely unlikely case, will need to take a deeper dive on the fix here because it doesn't seem trivial. We could check and not cancel the cte futures in the schedule() method but then it would need the info passed to the future

Reply:
I think I can add make all methods of CTEMaterializationTracker synchronized and remove all futures from the materializationFutures maps when the shedule() method resumes and cancels all pending futures [here](https://github.com/prestodb/presto/blob/85cc36af2cca5be4b15d17884f3f6a01bd107781/presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java#L517). Then it should be threadsafe.

Reply:
can we ignore the fact that it's been canceled?

Reply:
Since our map in CTEMaterializationTracker is storing the cancelled future, the [FixedSourceCountScheduler](https://github.com/prestodb/presto/blob/85cc36af2cca5be4b15d17884f3f6a01bd107781/presto-main/src/main/java/com/facebook/presto/execution/scheduler/FixedSourcePartitionedScheduler.java#L178) schedule method will return this cancelled future (which was cancelled by a previous loop of the scheduling process). I think that we need to make sure that the  FixedSourcePartitionedScheduler returns a fresh non cancelled future

Reply:
https://github.com/prestodb/presto/pull/24108/files#diff-bbcd68b946ac8590a0a1d1371f51e32687b0061d914689eb6fb3b852b69d3ac1R534 (main change clearing all futues)